### PR TITLE
SONAR-6547 Run one-time migration to clear the overloaded SQALE info on the Rules table

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import org.sonar.db.loadedtemplate.LoadedTemplateDto;
+import org.sonar.db.rule.RuleDto;
+import org.sonar.server.db.DbClient;
+
+import static org.sonar.db.loadedtemplate.LoadedTemplateDto.ONE_SHOT_TASK_TYPE;
+
+/**
+ * Clear the overloaded technical debt of rules when SQALE plugin is not installed.
+ *
+ * Should be removed after next LTS
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: https://jira.sonarsource.com/browse/SONAR-6547 is well commented, so it deserves to be linked here.

Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: "after next LTS" -> "after LTS 5.x" to be more precise.

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      registerTasks(session);
+      session.commit();
+    } finally {
+      dbClient.closeSession(session);
+    }
+  }
+
+  private void clearDebt(DbSession session) {
+    int countClearedRules = 0;
+    for (RuleDto rule : dbClient.deprecatedRuleDao().selectAll(session)) {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: I suggest to not use the deprecated dao but `dbClient.ruleDao()`. It requires to move `selectAll()`, which is a good thing for future developments.

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      if (isDebtOverridden(rule)) {
+        rule.setSubCharacteristicId(null);
+        rule.setRemediationFunction(null);
+        rule.setRemediationCoefficient(null);
+        rule.setRemediationOffset(null);
+        dbClient.deprecatedRuleDao().update(session, rule);
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: same here!

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+  private boolean isSqalePluginInstalled(DbSession session) {
+    return dbClient.propertiesDao().selectGlobalProperty(session, SQALE_LICENSE_PROPERTY) != null;
+  }
+
+  private boolean hasAlreadyBeenExecuted(DbSession session) {
+    return dbClient.loadedTemplateDao().countByTypeAndKey(ONE_SHOT_TASK_TYPE, TEMPLATE_KEY, session) > 0;
+  }
+
+  private void registerTasks(DbSession session) {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: I know that `loadedTemplate` is badly named, but consistency is more important. I suggest to rename the method `markAsExecuted()` so it does not introduce the term `task` and it's associate with the other method `hasAlreadyBeenExecuted`.

**Comment by simonbrandhof** [comment]
Some feedbacks about naming, else LGTM

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    }
+    if (countClearedRules > 0) {
+      LOG.warn("The SQALE model has been cleaned to remove useless data left over by previous migrations. The technical debt of {} rules was reset to their default values.",
+        countClearedRules);
+      LOG.warn("=> As a consequence, the overall technical debt of your projects might slightly evolve during the next analysis.");
+    }
+  }
+
+  private static boolean isDebtOverridden(RuleDto ruleDto) {
+    return ruleDto.getSubCharacteristicId() != null || ruleDto.getRemediationFunction() != null;
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: Lines 75 to 78 clean-up 4 columns, so they should all be tested here, even if you know it's not supposed to be possible.

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java (hunk 2 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: If I'm not using the deprecated DAO, how will it be possible to update the ES index ?

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java (hunk 1 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java (hunk 4 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java (hunk 5 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java (hunk 2 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/startup/ClearRulesOverloadedDebt.java]: good point!

**Comment by julienlancelot** [comment]
@simon, feedbacks applied here and inhttps://github.com/SonarSource/sonarqube/pull/536, back to you !

**Comment by simonbrandhof** [comment]
LGTM

**Pull request merged**